source/common/config/filesystem_subscription_impl.cc

```diff
@@ -77,5 +77,80 @@ void FilesystemSubscriptionImpl::refresh() {
+                                 validation_visitor, api) {}
+
+void FilesystemCollectionSubscriptionImpl::refresh() {
+  ENVOY_LOG(debug, "Filesystem collection config refresh for {}", path_);
+  stats_.update_attempt_.inc();
+  bool config_update_available = false;
+  envoy::service::discovery::v3::Resource resource_message;
+  try {
+    MessageUtil::loadFromFile(path_, resource_message, validation_visitor_, api_);
+    // Dynamically load the collection message.
+    const std::string collection_type =
+        std::string(TypeUtil::typeUrlToDescriptorFullName(resource_message.resource().type_url()));
+    const Protobuf::Descriptor* collection_descriptor =
+        Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(collection_type);
+    if (collection_descriptor == nullptr) {
+      throw EnvoyException(fmt::format("Unknown collection type {}", collection_type));
+    }
+    Protobuf::DynamicMessageFactory dmf;
+    ProtobufTypes::MessagePtr collection_message;
+    collection_message.reset(dmf.GetPrototype(collection_descriptor)->New());
+    MessageUtil::unpackTo(resource_message.resource(), *collection_message);
+    const auto* collection_entries_field_descriptor = collection_descriptor->field(0);
+    // Verify collection message type structure.
+    if (collection_entries_field_descriptor == nullptr ||
+        collection_entries_field_descriptor->type() != Protobuf::FieldDescriptor::TYPE_MESSAGE ||
+        collection_entries_field_descriptor->message_type()->full_name() !=
+            "udpa.core.v1.CollectionEntry" ||
+        !collection_entries_field_descriptor->is_repeated()) {
+      throw EnvoyException(fmt::format("Invalid structure for collection type {} in {}",
+                                       collection_type, resource_message.DebugString()));
+    }
+    const auto* reflection = collection_message->GetReflection();
+    const uint32_t num_entries =
+        reflection->FieldSize(*collection_message, collection_entries_field_descriptor);
+    DecodedResourcesWrapper decoded_resources;
+    for (uint32_t i = 0; i < num_entries; ++i) {
+      udpa::core::v1::CollectionEntry collection_entry;
+      collection_entry.MergeFrom(reflection->GetRepeatedMessage(
+          *collection_message, collection_entries_field_descriptor, i));
+      // TODO(htuch): implement indirect collection entries.
+      if (collection_entry.has_inline_entry()) {
+        decoded_resources.push_back(std::make_unique<DecodedResourceImpl>(
+            resource_decoder_, collection_entry.inline_entry()));
+      }
+    }
+    config_update_available = true;
+    callbacks_.onConfigUpdate(decoded_resources.refvec_, resource_message.version());
+    stats_.update_time_.set(DateUtil::nowToMilliseconds(api_.timeSource()));
+    stats_.version_.set(HashUtil::xxHash64(resource_message.version()));
+    stats_.version_text_.set(resource_message.version());
+    stats_.update_success_.inc();
```

Comment:
There is a decent amount of duplication between the non-collection refresh function above and this one. Is it possible to merge the implementations at all? Some of this is pretty nuanced and it would be nice to not have it duplicated in multiple places. This is just one example but there is more below.

Reply:
To some extent, yeah. I can refactor some parts in this PR. In future PRs, the collection implementation will actually own concrete filesystem subscriptions, since it will be taking the list collection and creating additional subscriptions for each entry.